### PR TITLE
Add kfp compiler

### PR DESCRIPTION
**Pull request by GeorgesLorre**
Untested -> looking for feedback on the structure not the content

### CODE REVIEW

File: src/fondant/pipeline.py [hunk 1]
```diff
@@ -303,240 +303,3 @@ def _validate_pipeline_definition(self, run_id: str):
...
-            The pipeline id.
-        """
-        return self.client.get_pipeline_id(pipeline_name)
-
-    def get_pipeline_version_ids(self, pipeline_id: str) -> t.List[str]:
-        """Function that returns the versions of a pipeline given a pipeline id."""
-        pipeline_versions = self.client.list_pipeline_versions(pipeline_id).versions
-        return [version.id for version in pipeline_versions]
-
-    def delete_pipeline(self, pipeline_name: str):
```
Comment (PhilippeMoussalli) [inline src/fondant/pipeline.py:451]: Where do you see all the other functionalities that are not related to compiling and running? user might still want to list pipeline runs, delete pipelines, ...

File: src/fondant/compiler.py [hunk 2]
```diff
@@ -183,3 +184,61 @@ def _generate_spec(self, pipeline: Pipeline, extra_volumes: list) -> dict:
...
             "services": services,
         }
+
+
+class KubeFlowCompiler(Compiler):
+    """Compiler that creates a Kubeflow pipeline spec from a pipeline."""
+
+    def _resolve_imports(self):
+        """Resolve imports for the Kubeflow compiler."""
+        try:
```
Comment (PhilippeMoussalli) [inline src/fondant/compiler.py:194]: I think you can just do 

if is_kfp_available():
    import kfp
    from kfp import dsl

since the function will check for those things for you

File: src/fondant/compiler.py [hunk 3]
```diff
@@ -183,3 +184,61 @@ def _generate_spec(self, pipeline: Pipeline, extra_volumes: list) -> dict:
...
+            from kfp import dsl
+        except ImportError:
+            raise ImportError(
+                "You need to install kfp to use the Kubeflow compiler, "
+                / "you can install it with `poetry install --extras kfp`"
+            )
+
+    def compile(
+        self,
+        pipeline: Pipeline,
```
Comment (PhilippeMoussalli) [inline src/fondant/compiler.py:205]: I think the `Pipeline` is the unifying interface between all runners right? if so, would add it to the abstract class

**Review by PhilippeMoussalli — COMMENTED** [review]
I like the overall structure and separation between the Compiler and Runner! 

I think there are still some things that don't fall quite between like Client related functionalities. Not the highest priority but we should figure out where those land

File: src/fondant/compiler.py (hunk 2 above)
Comment (RobbeSneyders) [reply src/fondant/compiler.py:194]: We should no longer need `is_kfp_available` with this approach. This is very nice change, since you'll no longer need to install kfp when using the local runner.

I think we can just move this to `__init__()` instead of this special function though?

File: src/fondant/compiler.py [hunk 4]
```diff
@@ -183,3 +184,61 @@ def _generate_spec(self, pipeline: Pipeline, extra_volumes: list) -> dict:
...
+
+    def _resolve_imports(self):
+        """Resolve imports for the Kubeflow compiler."""
+        try:
+            import kfp
+            from kfp import dsl
+        except ImportError:
+            raise ImportError(
+                "You need to install kfp to use the Kubeflow compiler, "
+                / "you can install it with `poetry install --extras kfp`"
```
Comment (RobbeSneyders) [inline src/fondant/compiler.py:200]:
```suggestion
                / "you can install it with `pip install fondant[kfp]`"
```

`poetry install` is only for fondant developers. Users can also use poetry, but then they'll install their own project and need to indicate the fondant extras with the fondant dependency in their `pyproject.toml`.

File: src/fondant/runner.py [hunk 5]
```diff
@@ -26,3 +29,30 @@ def run(cls, input_spec: str, *args, **kwargs):
...
+class KubeflowRunner(Runner):
+    @abstractmethod
+    def _resolve_imports(self):
+        """Resolve imports for the Kubeflow compiler."""
+        try:
+            import kfp
+        except ImportError:
+            raise ImportError(
+                "You need to install kfp to use the Kubeflow compiler, "
+                / "you can install it with `poetry install --extras kfp`"
```
Comment (RobbeSneyders) [inline src/fondant/runner.py:43]:
```suggestion
                "You need to install kfp to use the Kubeflow runner, "
                / "you can install it with `pip install fondant[kfp]`"
```

File: src/fondant/pipeline.py (hunk 1 above)
Comment (RobbeSneyders) [reply src/fondant/pipeline.py:451]: Not sure if we need to wrap this functionality for every platform we want to run on. At least at this point in time, I would be ok with removing it and having the user work with the underlying platform directly.

File: src/fondant/runner.py [hunk 6]
```diff
@@ -26,3 +29,30 @@ def run(cls, input_spec: str, *args, **kwargs):
...
+            )
+
+    def run(cls, input_spec: str, host: str, *args, **kwargs):
+        """Run a kubeflow pipeline."""
+        cls._resolve_imports()
+        client = kfp.Client(host=host)
```
Comment (RobbeSneyders) [inline src/fondant/runner.py:49]: This probably doesn't change a lot to the current functionality, but it would make more sense to me to instantiate this in `__init__()`.

**Review by RobbeSneyders — COMMENTED** [review]
Thanks @GeorgesLorre!

The lines along you split the behavior look good to me!

File: src/fondant/compiler.py (hunk 4 above)
Comment (PhilippeMoussalli) [reply src/fondant/compiler.py:200]: https://github.com/ml6team/fondant/blob/86867a5e777c93a7ca0d982f263d66c38abc3e8f/src/fondant/import_utils.py#L24C1-L25C1

File: src/fondant/compiler.py (hunk 2 above)
Comment (PhilippeMoussalli) [reply src/fondant/compiler.py:194]: Created a PR to remove obsolete packages

https://github.com/ml6team/fondant/pull/293

File: src/fondant/pipeline.py (hunk 1 above)
Comment (PhilippeMoussalli) [reply src/fondant/pipeline.py:451]: makes sense, we'll re-add it later if we notice that it's needed

File: src/fondant/compiler.py [hunk 7]
```diff
@@ -207,3 +207,108 @@ def _generate_spec(
...
+
+    def _resolve_imports(self):
+        """Resolve imports for the Kubeflow compiler."""
+        try:
+            import kfp
+
+            self.kfp = kfp
+        except ImportError:
+            msg = """You need to install kfp to use the Kubeflow compiler,\n
+                     you can install it with `pip install --extras kfp`"""
```
Comment (RobbeSneyders) [inline src/fondant/compiler.py:226]:
```suggestion
                     you can install it with `pip install fondant[pipelines]`"""
```

Although we probably want to split this up for the different runners so it becomes fondant[kfp].

File: src/fondant/compiler.py [hunk 8]
```diff
@@ -207,3 +207,108 @@ def _generate_spec(
...
+    ) -> None:
+        """Compile a pipeline to Kubeflow pipeline spec and save it to a specified output path.
+
+        Args:
+            pipeline: the pipeline to compile
+            output_path: the path where to save the Kubeflow pipeline spec
+        """
+        self.pipeline = pipeline
+        self.pipeline.sort_graph()
+        self.pipeline._validate_pipeline_definition("{{workflow.name}}")
```
Comment (RobbeSneyders) [inline src/fondant/compiler.py:244]: This should be done in the `DockerCompiler` as well, no?

File: src/fondant/pipeline.py [hunk 9]
```diff
@@ -18,8 +17,6 @@
 from fondant.schema import validate_partition_number, validate_partition_size
 
 if is_kfp_available():
```
Comment (RobbeSneyders) [inline src/fondant/pipeline.py:19]: Ideally we can get rid of this and make this file completely runner independent.

File: src/fondant/runner.py [hunk 10]
```diff
@@ -26,3 +29,34 @@ def run(self, input_spec: str, *args, **kwargs):
...
+
+class KubeflowRunner(Runner):
+    def __init__(self):
+        self._resolve_imports()
+
+    @abstractmethod
+    def _resolve_imports(self):
+        """Resolve imports for the Kubeflow compiler."""
+        try:
+            global kfp
```
Comment (RobbeSneyders) [inline src/fondant/runner.py:42]: Can we assign to `self.kfp` here as in the compiler?

File: src/fondant/runner.py [hunk 11]
```diff
@@ -26,3 +29,34 @@ def run(self, input_spec: str, *args, **kwargs):
...
+        # client = kfp.Client(host=host)
+        # # TODO add logic to see if pipeline exists
+        # pipeline_spec = client.run_pipeline(
+        #     experiment_id=experiment.id,
+        #     job_name=run_name,
+        #     pipeline_package_path=pipeline.package_path,
+        # )
+
+        # pipeline_url = f"{self.host}/#/runs/details/{pipeline_spec.id}"
+        # logger.info(f"Pipeline is running at: {pipeline_url}")
```
Comment (RobbeSneyders) [inline src/fondant/runner.py:62]: Is there a reason this is still commented out?

File: tests/test_pipeline.py [hunk 12]
```diff
@@ -169,12 +165,12 @@ def test_invalid_pipeline_dependencies(
...
+        ("example_3", ["first_component", "second_component"]),
     ],
 )
 def test_invalid_pipeline_compilation(
     default_pipeline_args,
     invalid_pipeline_example,
-    tmp_path,
 ):
     """
     Test that an InvalidPipelineDefinition exception is raised when attempting to compile
```
Comment (RobbeSneyders) [inline tests/test_pipeline.py:176]: You're no longer compiling in the test :)

**Review by RobbeSneyders — COMMENTED** [review]
Thanks @GeorgesLorre!

The general direction looks good. Left some comments.

File: src/fondant/components
Comment (RobbeSneyders) [inline src/fondant/components:1]: This should not be removed

File: src/fondant/runner.py (hunk 11 above)
Comment (GeorgesLorre) [reply src/fondant/runner.py:62]: should not have committed this, will be taken up in a separate PR addressing the kfp runner

File: src/fondant/pipeline.py (hunk 9 above)
Comment (PhilippeMoussalli) [reply src/fondant/pipeline.py:19]: This is only needed for setting `p_volumes` at the `componentOp` level, that was only needed when we were initially planning to download the Laion dataset locally to use it. I think we can remove both `ephemeral_storage_size` and `p_volumes` from the ComponenOp. We can always re-add them later if we notice they're needed

**Review by PhilippeMoussalli — COMMENTED** [review]
Thanks Georges! Looks much cleaner already ✨

**Comment by GeorgesLorre** [comment]
@RobbeSneyders updated!

File: src/fondant/compiler.py [hunk 13]
```diff
@@ -207,3 +207,108 @@ def _generate_spec(
...
+
+    def __init__(self):
+        self._resolve_imports()
+
+    def _resolve_imports(self):
+        """Resolve imports for the Kubeflow compiler."""
+        try:
+            import kfp
+
+            self.kfp = kfp
```
[the remark this answers is not in the document]
Comment (GeorgesLorre) [reply src/fondant/compiler.py:226]: it avoids having to use a global, this is cleaner imo

**Comment by GeorgesLorre** [comment]
@PhilippeMoussalli Ok I'll remove the `ephemeral_storage_size` and `p_volumes` in a new PR. This would indeed be cleaner.

**Review by RobbeSneyders — APPROVED** [review]
Thanks! Can you just update the error message for the kfp import before merging?
